File: bleep-core/src/scala/bleep/BleepScript.scala
package bleep

import bleep.rewrites.BuildRewrite

abstract class BleepScript(val scriptName: String) {
  // can override this in subclass
  val rewrites: List[BuildRewrite] = Nil

  def main(args: Array[String]): Unit = {
    val (preOpts, restArgs) = PreBootstrapOpts.parse(args.toList)
    bootstrap.forScript(scriptName, preOpts.toLoggingOpts, rewrites)((started, commands) => run(started, commands, restArgs))
  }

  def run(started: Started, commands: Commands, args: List[String]): Unit
}

File: bleepscript/src/main/java/bleepscript/BleepScript.java
package bleepscript;

import java.util.List;
import java.util.Objects;

/**
 * Base class for user-authored bleep scripts written in Java.
 *
 * <p>Typical usage:
 *
 * <pre>{@code
 * public final class MyScript extends BleepScript {
 *   public MyScript() { super("my-script"); }
 *
 *   @Override
 *   public void run(Started started, Commands commands, List<String> args) {
 *     started.logger().info("hello");
 *     commands.compile(List.of(CrossProjectName.of("my-project")));
 *   }
 *
 *   public static void main(String[] args) {
 *     new MyScript().bootstrap(args);
 *   }
 * }
 * }</pre>
 */
public abstract class BleepScript {
  private final String name;

  protected BleepScript(String name) {
    this.name = Objects.requireNonNull(name, "name");
  }

  public final String scriptName() {
    return name;
  }

  public abstract void run(Started started, Commands commands, List<String> args);

  public final void bootstrap(String[] args) {
    BleepscriptServices.Holder.INSTANCE.forScript(name, args, this);
  }
}
